R/g.analyse.perday.R

```diff
@@ -320,23 +324,50 @@ g.analyse.perday = function(ndays, firstmidnighti, time, nfeatures,
             for (mi in 2:ncol(metashort)) { #run through metrics (for features based on single metrics)
               NRV = length(which(is.na(as.numeric(as.matrix(vari[,mi]))) == FALSE))
               varnum = as.numeric(as.matrix(vari[,mi])) #varnum is one column of vari
+              vari2 = vari #vari2 is used in identify_levels
+              if (isTRUE(params_cleaning[["part2ExcludeNonwear"]])) {
+                varnum[which(val != 0)] = NA
+              } 
               # # if this is the first or last day and it has more than includedaycrit number of days then expand it
               # Comment out the following 10 lines if you want to include only the actual data
-              if (NRV != length(averageday[, (mi - 1)])) {
+              if (NRV < length(averageday[, (mi - 1)])) { # modified to avoid imputing time in the daylight saving days (25-h long, then they meet the condition NRV != nrow(averageday))
                 difference = NRV - length(averageday[, (mi - 1)])
                 if (di == 1) {
-                  varnum = c(averageday[1:abs(difference), (mi - 1)], varnum)
-                  # readjust anwi indices in case that varnum has been imputed
-                  if (max(anwi_t1) < length(varnum)) { # since GGIR always calculates full window, max(anwi_t1) should always equals length(varnum)
-                    anwi_t0 = anwi_t0 + abs(difference)
-                    anwi_t1 = anwi_t1 + abs(difference)
-                    # then, we reset the minimum anwi_t0 to 1 to consider the imputed varnum
-                    anwi_t0[which(anwi_t0 == min(anwi_t0))] = 1
+                  # varnum = c(averageday[1:abs(difference), (mi - 1)], varnum)
+                  varnum = c(rep(NA, abs(difference)), varnum) # 2022-11-1: if day <24h, then the total time analised is also <24h 
```

Comment:
We cannot add NA values here, because further down these NAs will be converted to zero and then treated as if they are a good imputation of the missing data. In other words if a recording starts at 5pm this code will assume that the person slept until 5pm.  Given that averageday is not informative when the accelerometer is not worn during the night, I think the only solution is to omit the entire first and last day of the recording and keep this part of the code as it was.

Reply:
See responses above, I think the places where I converted NAs to zero are not affecting the estimates, but making them consistent with full days (24h). In the files I have tried this I get the expected output, but I am happy to do more tests if you can think of any scenario that might be affected for this. Also, if you think we should limit the estimates of part2 to only complete days, then we can add a condition for this [here](https://github.com/wadpac/GGIR/blob/master/R/g.analyse.perday.R#L330-L335:~:text=if%20(nvalidhours%20%3E%3D%20includedaycrit)%20%7B).

In the master branch, this time is replaced by `averageday`, which makes that the `Nhours` is wrongly defined as 24h even if recording started at (e.g.) 6am in the first day, and can produce that some imputed time is classified as MVPA (even if the accelerometer was not recording data).